mater/lib/src/lib.rs

```diff
@@ -18,16 +18,178 @@ mod unixfs;
+};
+use tokio::io::{AsyncReadExt, AsyncSeek, AsyncSeekExt, AsyncWriteExt};
 pub use v1::{Header as CarV1Header, Reader as CarV1Reader, Writer as CarV1Writer};
 pub use v2::{
     verify_cid, Characteristics, Header as CarV2Header, Index, IndexEntry, IndexSorted,
     MultihashIndexSorted, Reader as CarV2Reader, SingleWidthIndex, Writer as CarV2Writer,
 };
 
+/// Represents the location and size of a block in the CAR file.
+pub struct BlockLocation {
+    /// The byte offset in the CAR file where the block starts.
+    pub offset: u64,
+    /// The size (in bytes) of the block.
+    pub size: u64,
+}
+
+/// A simple blockstore backed by a CAR file and its index.
+pub struct CarBlockStore<R> {
+    reader: R,
+    /// Mapping from CID to block location.
+    pub index: HashMap<Cid, BlockLocation>,
+}
+
+impl<R> CarBlockStore<R>
+where
+    R: AsyncSeek + AsyncReadExt + Unpin,
+{
+    /// Loads the CARv2 index at the specified `index_offset` and constructs a
+    /// mapping from [`Cid`] to [`BlockLocation`] (offset, size).
+    ///
+    /// # Index Format
+    ///
+    /// > **Note on Spec Compliance**
+    /// > The [CARv2 specification](https://ipld.io/specs/transport/car/carv2/#index-format) states
+    /// > that the first byte(s) at the index offset contain a varint-based multicodec code to
+    /// > identify the index format (e.g., `IndexSorted`, `MultihashIndexSorted`), and the remaining
+    /// > bytes follow that format’s rules.
+    /// >
+    /// > **Here**, we do **not** parse any varint-based format code; instead we assume a single,
+    /// > fixed layout for each index entry:
+    /// >
+    /// > ```text
+    /// > [CID length (u8)] [CID bytes] [offset (u64)] [size (u64)]
+    /// > ```
+    /// >
+    /// > If you need to handle multiple official CARv2 index types, you must first read the varint
+    /// > code and dispatch to the appropriate parser. As of now, this function assumes a single
+    /// > custom format is sufficient for our use case.
+    ///
+    /// We read entries in a loop until reaching EOF, at which point we assume
+    /// there are no more index entries. Any I/O errors other than EOF cause an
+    /// immediate return with an error.
+    ///
+    /// # Assumptions
+    ///
+    /// - The CAR header has already been processed, and `index_offset`
+    ///   points to the start of the index data.
+    /// - `cid_len` fits within a single `u8`; larger CIDs are not expected.
+    ///   If your use case requires arbitrary-length CIDs or multiple index formats,
+    ///   you’ll need to extend this function accordingly.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::IoError`] if any non-EOF I/O error occurs.
+    /// - Returns [`Error::InvalidCid`] if CID decoding fails.
+    /// - Returns [`Error::BlockNotFound`] if a requested block is missing (in `get_block`).
+    pub async fn load_index(
+        mut reader: R,
+        index_offset: u64,
```

Comment:
According to the spec, if `index_offset` is 0, it means that the index is not there. Are we handling that? (https://ipld.io/specs/transport/car/carv2/#index-payload)